design/src/rfcs/rfc0022_refine_builder.md

```diff
@@ -0,0 +1,556 @@
+
+Given the above, we think that the impact of a runtime error is low enough to be worth exploring designs that do not guarantee compile-safety for the builder API[^further-dev-productivity-improvements].  
+
+### Providing clear feedback
+
+Moving from a compile-time error to a runtime error does not require extensive refactoring.  
+The definition of `PokemonServiceBuilder` goes from:
+
+```rust
+pub struct PokemonServiceBuilder<
+    Op1,
+    Op2,
+    Op3,
+    Op4,
+    Op5,
+    Op6,
+    Exts1 = (),
+    Exts2 = (),
+    Exts3 = (),
+    Exts4 = (),
+    Exts5 = (),
+    Exts6 = (),
+    Pl = aws_smithy_http_server::plugin::IdentityPlugin,
+> {
+    check_health: Op1,
+    do_nothing: Op2,
+    get_pokemon_species: Op3,
+    get_server_statistics: Op4,
+    capture_pokemon: Op5,
+    get_storage: Op6,
+    #[allow(unused_parens)]
+    _exts: std::marker::PhantomData<(Exts1, Exts2, Exts3, Exts4, Exts5, Exts6)>,
+    plugin: Pl,
+}
+```
+
+to:
+
+```rust
+pub struct PokemonServiceBuilder<
+    Op1,
+    Op2,
+    Op3,
+    Op4,
+    Op5,
+    Op6,
+    Exts1 = (),
+    Exts2 = (),
+    Exts3 = (),
+    Exts4 = (),
+    Exts5 = (),
+    Exts6 = (),
+    Pl = aws_smithy_http_server::plugin::IdentityPlugin,
+> {
+    check_health: Option<Op1>,
+    do_nothing: Option<Op2>,
+    get_pokemon_species: Option<Op3>,
+    get_server_statistics: Option<Op4>,
+    capture_pokemon: Option<Op5>,
+    get_storage: Option<Op6>,
+    #[allow(unused_parens)]
+    _exts: std::marker::PhantomData<(Exts1, Exts2, Exts3, Exts4, Exts5, Exts6)>,
+    plugin: Pl,
+}
+```
+
+All operation fields are now `Option`-wrapped.  
+We introduce a new `MissingOperationsError` error to hold the names of the missing operations and their respective setter methods:
+
+```rust
+#[derive(Debug)]
+pub struct MissingOperationsError {
+    service_name: &'static str,
+    operation_names2setter_methods: HashMap<&'static str, &'static str>,
+}
+
+impl Display for MissingOperationsError { /* */ }
+impl std::error::Error for MissingOperationsError {}
+```
+
+which is then used in `build` as error type _(not shown here for brevity)_.  
+We can now try again to stray away from the happy path by forgetting to register a handler for the `CheckHealth` operation.  
+The code compiles just fine this time, but the application fails when launched via `cargo run`:
+
+```text
+<timestamp> ERROR pokemon_service: You must specify a handler for all operations attached to the `Pokemon` service.
+We are missing handlers for the following operations:
+- com.aws.example#CheckHealth
+
+Use the dedicated methods on `PokemonServiceBuilder` to register the missing handlers:
+- PokemonServiceBuilder::check_health
```

Comment:
Beginners to rust may not know that this is telling them to use `builder.check_health()`, maybe we should clarify that.